cranelift/codegen/src/isa/aarch64/inst.isle

```diff
@@ -2593,3 +2632,134 @@
+      (let ((dst WritableReg (temp_writable_reg $F64))
+            (_ Unit (emit (MInst.FpuMove64 dst x)))
+            (tmp Reg (fpu_rri (fpu_op_ri_ushr (ty_bits ty) (max_shift ty)) y))
+            (_ Unit (emit (MInst.FpuRRI (fpu_op_ri_sli (ty_bits ty) (max_shift ty)) dst tmp))))
+       dst))
+(rule (fcopy_sign x y ty @ (multi_lane _ _))
+      (let ((dst WritableReg (temp_writable_reg $I8X16))
+            (_ Unit (emit (MInst.FpuMove128 dst x)))
+            (tmp Reg (vec_shift_imm (VecShiftImmOp.Ushr) (max_shift (lane_type ty)) y (vector_size ty)))
+            (_ Unit (emit (MInst.VecShiftImmMod (VecShiftImmModOp.Sli) dst tmp (vector_size ty) (max_shift (lane_type ty))))))
+       dst))
+
+;; Helpers for generating `MInst.FpuToInt` instructions.
+
+(decl fpu_to_int_nan_check () ConsumesFlags)
+(rule (fpu_to_int_nan_check)
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond (Cond.Vs))
+                    (trap_code_bad_conversion_to_integer))))
+
+;; Emits the appropriate flag-reading op for an underflow check,
+;; accepting a boolean (whether the type is signed), input type,
+;; and output type.
+(decl fpu_to_int_underflow_check (bool Type Type) ConsumesFlags)
+(rule (fpu_to_int_underflow_check $true $F32 (fits_in_16 out_ty))
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond
+                  (Cond.Le))
+                  (trap_code_integer_overflow))))
+(rule (fpu_to_int_underflow_check $true $F64 (fits_in_32 out_ty))
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond
+                  (Cond.Le))
+                  (trap_code_integer_overflow))))
+(rule -1 (fpu_to_int_underflow_check $true _in_ty _out_ty)
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond
+                  (Cond.Lt))
+                  (trap_code_integer_overflow))))
+(rule (fpu_to_int_underflow_check $false _in_ty _out_ty)
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond
+                  (Cond.Le))
+                  (trap_code_integer_overflow))))
+
+(decl fpu_to_int_overflow_check () ConsumesFlags)
+(rule (fpu_to_int_overflow_check)
+      (ConsumesFlags.ConsumesFlagsSideEffect
+      (MInst.TrapIf (cond_br_cond
+                  (Cond.Ge))
+                  (trap_code_integer_overflow))))
+
+;; Emits the appropriate instruction sequence to convert a
+;; floating-point value to an integer, trapping if the value
+;; is a NaN or does not fit in the target type.
+;; Accepts the specific conversion op, the source register,
+;; whether the input is signed, and finally the input and output
+;; types.
+(decl fpu_to_int_cvt (FpuToIntOp Reg bool Type Type) Reg)
+(rule (fpu_to_int_cvt op src signed in_ty out_ty)
+      (let ((size ScalarSize (scalar_size in_ty))
+            (in_bits u8 (ty_bits in_ty))
+            (out_bits u8 (ty_bits out_ty))
+            (_ InstOutput (side_effect (with_flags_side_effect
```

Comment:
I wonder, would it make more sense to build an abstraction here that doesn't require the explicit invoke-for-side-effect-and-discard pattern, perhaps a helper that takes `src`, checks if NaN and traps if so, and then returns `src`? Then we could do something like `(final_expression (check_nan_and_trap src) ...)` in the body of the `let`. Likewise for `min` and `max` (underflow and overflow checks).